File: runtime/java/units/wasi/errno_fs.java
// Filesystem-only errno codes: kept out of the always-bundled
// wasi/_class prelude so a stdio-only WASI module (no path_* / fs-only fd_*
// imports) doesn't carry them. The base codes (BADF/INVAL/IO/SPIPE) live in
// _class since stdio needs them too.
static final int WASI_ACCES = 2;
static final int WASI_EXIST = 20;
static final int WASI_ISDIR = 31;
static final int WASI_LOOP = 32;
static final int WASI_NAMETOOLONG = 37;
static final int WASI_NOENT = 44;
static final int WASI_NOTDIR = 54;
static final int WASI_NOTEMPTY = 55;
static final int WASI_PERM = 63;
// WASI_NOTCAPABLE (76) lives in the always-bundled wasi/_class prelude, since
// the rights model enforces it from stdio-core fd_* units too.

// One host-error-to-WASI-errno mapping shared by every filesystem syscall, so
// the same host error never maps to different codes depending on which syscall
// raised it. Java's NIO raises typed subclasses of IOException, so match on
// those; everything else falls back to EIO. Note the honest gaps vs
// the Go/Python backends, which read the raw errno: Java exposes no distinct
// exception for EISDIR, ELOOP, or ENAMETOOLONG at open/stat time, so those
// host conditions surface as EIO here unless a syscall detects them itself
// (path_unlink_file/path_remove_directory pre-check for EISDIR/ENOTDIR).
int fs_errno(java.io.IOException e) {
    if (e instanceof java.nio.file.NoSuchFileException) {
        return WASI_NOENT;
    }
    if (e instanceof java.nio.file.FileAlreadyExistsException) {
        return WASI_EXIST;
    }
    if (e instanceof java.nio.file.AccessDeniedException) {
        return WASI_ACCES;
    }
    if (e instanceof java.nio.file.DirectoryNotEmptyException) {
        return WASI_NOTEMPTY;
    }
    if (e instanceof java.nio.file.NotDirectoryException) {
        return WASI_NOTDIR;
    }
    return WASI_IO;
}

File: runtime/go/units/wasi/errno_fs.go
// Filesystem-only errno codes: kept out of the always-bundled wasi/_class prelude so a stdio-only WASI module (no path_* / fs-only fd_* imports) doesn't carry them.
// The base codes (BADF/INVAL/IO/SPIPE) live in
// _class since stdio needs them too.
const (
    wasiAcces       uint32 = 2
    wasiExist       uint32 = 20
    wasiIsdir       uint32 = 31
    wasiLoop        uint32 = 32
    wasiNametoolong uint32 = 37
    wasiNoent       uint32 = 44
    wasiNotdir      uint32 = 54
    wasiNotempty    uint32 = 55
    wasiPerm        uint32 = 63
    // wasiNotcapable (76) lives in wasi/_class: the per-fd rights model needs it even in stdio-only modules that don't bundle these fs-only codes.
)

// One host-error-to-WASI-errno mapping shared by every filesystem syscall, so the same host error never maps to different codes depending on which syscall raised it.
// Go wraps the raw errno in *fs.PathError/*os.LinkError, so unwrap it with errors.As.
func (w *WASI) fs_errno(err error) uint32 {
    var errno syscall.Errno
    if errors.As(err, &errno) {
        switch errno {
        case syscall.EACCES:
            return wasiAcces
        case syscall.EBADF:
            return wasiBadf
        case syscall.EEXIST:
            return wasiExist
        case syscall.EINVAL:
            return wasiInval
        case syscall.EISDIR:
            return wasiIsdir
        case syscall.ELOOP:
            return wasiLoop
        case syscall.ENAMETOOLONG:
            return wasiNametoolong
        case syscall.ENOENT:
            return wasiNoent
        case syscall.ENOTDIR:
            return wasiNotdir
        case syscall.ENOTEMPTY:
            return wasiNotempty
        case syscall.EPERM:
            return wasiPerm
        case syscall.ESPIPE:
            return wasiSpipe
        }
    }
    return wasiIo
}
